### PR TITLE
[typescript-node] Promise-based Authentication interface

### LINKED ISSUE #3407: [REQ] [typescript-node] Promise-based Authentication interface
### Is your feature request related to a problem? Please describe.

The current TypeScript Node [`Authentication`](https://github.com/OpenAPITools/openapi-generator/blob/21e29d7177/modules/openapi-generator/src/main/resources/typescript-node/models.mustache#L165) interface for applying authentication settings (`.applyToRequest`) is synchronous.  It would be useful to support a Promise-based asynchronous API also.  For example, to fetch a refreshed authentication token.

https://github.com/kubernetes-client/javascript/issues/294#issuecomment-510323971 is one example that would use this feature.  

## Describe the solution you'd like

Update `Authentication.applyToRequest` to return `Promise<void> | void`.

## Describe alternatives you've considered

1. "`spawnSync`-style hacks", like serializing arguments and passing them to an script and parsing the result
2. Breaking change to `Authentication.applyToRequest` that forces all implementers to return a `Promise<void>`

### PR DESCRIPTION
**Pull request by silasbw**
### PR checklist

- [x] Read the [contribution guidelines](https://github.com/openapitools/openapi-generator/blob/master/CONTRIBUTING.md).
- [x] Ran the shell script under `./bin/` to update Petstore sample so that CIs can verify the change. (For instance, only need to run `./bin/{LANG}-petstore.sh`, `./bin/openapi3/{LANG}-petstore.sh` if updating the {LANG} (e.g. php, ruby, python, etc) code generator or {LANG} client's mustache templates). Windows batch files can be found in `.\bin\windows\`. If contributing template-only or documentation-only changes which will change sample output, be sure to [build the project](https://github.com/OpenAPITools/openapi-generator#14---build-projects) first.
- [ ] Filed the PR against the [correct branch](https://github.com/OpenAPITools/openapi-generator/wiki/Git-Branches): `master`, `4.1.x`, `5.0.x`. Default: `master`.
- [x] Copied the [technical committee](https://github.com/openapitools/openapi-generator/#62---openapi-generator-technical-committee) to review the pull request if your PR is targeting a particular programming language.

### Description of the PR

Example implementation for https://github.com/OpenAPITools/openapi-generator/issues/3407.

fixes #3407

### CODE REVIEW

**Comment by silasbw** [comment]
cc @TiFu (2017/07) @taxpon (2017/07) @sebastianhaas (2017/07) @kenisteward (2017/07) @Vrolijkx (2017/09) @macjohnny (2018/01) @nicokoenig (2018/09) @topce (2018/10) @akehir (2019/07)

**Comment by silasbw** [comment]
:wave: @wing328

File: modules/openapi-generator/src/main/resources/typescript-node/api-single.mustache [hunk 1]
```diff
@@ -197,33 +197,35 @@ export class {{classname}} {
 {{/bodyParam}}
         };
 
+        let authenticationPromise = Promise.resolve();
```
Comment (macjohnny) [inline modules/openapi-generator/src/main/resources/typescript-node/api-single.mustache]: what is the use of this promise?

File: modules/openapi-generator/src/main/resources/typescript-node/api-single.mustache [hunk 2]
```diff
@@ -197,33 +197,35 @@ export class {{classname}} {
...
-            } else {
-                localVarRequestOptions.form = localVarFormParams;
-            }
-        }
-        return new Promise<{ response: http.{{#supportsES6}}IncomingMessage{{/supportsES6}}{{^supportsES6}}ClientResponse{{/supportsES6}}; {{#returnType}}body: {{{returnType}}}; {{/returnType}}{{^returnType}}body?: any; {{/returnType}} }>((resolve, reject) => {
-            localVarRequest(localVarRequestOptions, (error, response, body) => {
-                if (error) {
-                    reject(error);
+        authenticationPromise = authenticationPromise.then(() => this.authentications.default.applyToRequest(localVarRequestOptions));
+        return authenticationPromise.then(() => {
```
Comment (macjohnny) [inline modules/openapi-generator/src/main/resources/typescript-node/api-single.mustache]: this return type doesn't match the method signature, does it?

File: modules/openapi-generator/src/main/resources/typescript-node/api-single.mustache (hunk 1 above)
Comment (silasbw) [reply modules/openapi-generator/src/main/resources/typescript-node/api-single.mustache]: To unconditionally setup the serial promise chain below.  It avoids `=== null` style checks.

An alternative I considered was executing the promises concurrently, storing them in an array, and using `Promise.all`.  That would change the semantics the the current synchronous and serial behavior.

File: modules/openapi-generator/src/main/resources/typescript-node/api-single.mustache (hunk 2 above)
Comment (silasbw) [reply modules/openapi-generator/src/main/resources/typescript-node/api-single.mustache]: I think it does and I empirically verified (i.e., compiled the petstore ts files).

Any suggestions on how I could make that more obvious?

File: modules/openapi-generator/src/main/resources/typescript-node/api-single.mustache (hunk 1 above)
Comment (macjohnny) [reply modules/openapi-generator/src/main/resources/typescript-node/api-single.mustache]: ah, ok, now I get it, the shorthand `return this.authentications.default.applyToRequest(localVarRequestOptions)` is the relevant part here.

File: modules/openapi-generator/src/main/resources/typescript-node/api-single.mustache (hunk 2 above)
Comment (macjohnny) [reply modules/openapi-generator/src/main/resources/typescript-node/api-single.mustache]: I missed the promise chaining, returning the response promise in the end

**Comment by macjohnny** [comment]
Yes, this should be part of 4.1.0

**Comment by wing328** [comment]
@silasbw thanks for the PR, which has been included in the 4.1.0 release: https://twitter.com/oas_generator/status/1160000504455319553

**Pull request merged**